src/Mapper.cpp

```diff
@@ -1346,3 +1346,223 @@ map<int, CGRANode*>* Mapper::getReorderPath(map<CGRANode*, int>* t_path) {
+        }
+
+        /*cout<<"Printing references read from increMapInput.json"<<endl;
+        for(map<DFGNode*, CGRANode*>::iterator iter = refMapRes.begin(); iter != refMapRes.end(); ++iter){
+                cout<<iter->first->getOpcodeName()<<iter->first->getID()<<"->Tile"<<iter->second->getID()<<", level="<<CGRANodeID2Level[iter->second->getID()]<<endl;
+        }
+        cout<<"Done..."<<endl;*/
+
+        return 0;
+}
+
+// generates variables for incremental mapping
+void Mapper::sortAllocTilesByLevel(CGRA* t_cgra){
+        map<int, vector<CGRANode*>> FanIO_CGRANodes;
+        vector<int> FanIOs;
+        int numTiles = 0;
+        for (int i=0; i<t_cgra->getRows(); ++i) {
+                for (int j=0; j<t_cgra->getColumns(); ++j) {
+
+                        // record the number of FanIO for each tile. 
+                        CGRANode* currentCGRANode = t_cgra->nodes[i][j];
+                        if(currentCGRANode->isDisabled()) continue; // only record tiles that have DFGNodes mapped
+                        int FanIO = max(currentCGRANode->getInLinks()->size(), currentCGRANode->getOutLinks()->size());
+                        FanIO_CGRANodes[FanIO].push_back(currentCGRANode);
+
+                        // record FanIO in the list if it appears for the first time
+                        if (find(FanIOs.begin(), FanIOs.end(), FanIO) == FanIOs.end()){
+                                FanIOs.push_back(FanIO);
+                        }
+
+                        numTiles++;
+                }
+        }
+        std::sort(FanIOs.rbegin(), FanIOs.rend()); // sort FanIOs from big to small
+
+        //cout<<"Generating vairable CGRANodes_sortedByLevel"<<endl;
+        CGRANodes_sortedByLevel.clear();
+        for(int level = 0; level < FanIOs.size(); level++){
+                int FanIO = FanIOs[level];
+                vector<CGRANode*> tiles = FanIO_CGRANodes[FanIO];
+                // for(auto tile:tiles) cout<<"Level"<<level<<": Tile"<<tile->getID()<<endl;
+                CGRANodes_sortedByLevel.push_back(tiles);
+        }
+        cout<<"Done..."<<endl;
+}
+
+// generates the placement recommendation list for current DFGNode by referencing its placement in the former mapping results
+// Two principles: RPT & MBO
+list<CGRANode*> Mapper::placementGen(CGRA* t_cgra,  DFGNode* t_dfgNode){
+        list<CGRANode*> placementRecommList;
+        CGRANode* refCGRANode = refMapRes[t_dfgNode];
+        list<DFGNode*>* predNodes = t_dfgNode->getPredNodes();
+        int refLevel = CGRANodeID2Level[refCGRANode->getID()];
+        int level = refLevel;
+        int maxLevel = CGRANodes_sortedByLevel.size() - 1;
+        cout<<t_dfgNode->getOpcodeName()<<t_dfgNode->getID()<<" is mapped to Tile "<<refCGRANode->getID()<<" in the referenced mapping results, refLevel="<<refLevel<<endl;
+
+        int initLevel = level;
+        while(true){
+                map<int, vector<CGRANode*>> bypassNums_CGRANode;
```

Comment:
Plz comment on what is this map used for.

Reply:
map<int, vector<CGRANode*>> bypassNums_CGRANode is used for sorting the CGRANodes with the number of bypass operations required to communicate with its predecessors.

Reply:
Has been corrected